### DISCUSSION TITLE
UpsertMany does not update the table with the new data.

### DISCUSSION THREAD

**Discussion by hugocruzlfc** [mikro-orm/mikro-orm#5926] [General]
### Describe the bug

I am trying to insert or update an array of types of my entity. I'm trying to use upsertMany, so far it inserts but never updates.

### Reproduction

```ts
@Injectable()
export class BaggageSearchRepository
	extends RepositoryImplementation<BaggageSearchModel>
	implements IBaggagesSearchRepository
{
	constructor(
		@InjectRepository(BaggageSearchEntity)
		private readonly baggageSearchRepository: EntityRepository<BaggageSearchEntity>,
	) {
		super();
	}



	protected toPersistence(
		baggageSearchModel: BaggageSearchModel,
	): RequiredEntityData<BaggageSearchEntity> {
		return {
			baggageSearchId: baggageSearchModel.baggageSearchId.value,
			bagTag: baggageSearchModel.bagTag.value,
			flightId: baggageSearchModel.flightId.value,
			status: baggageSearchModel.status.value,
		};
	}

	async upsertMany(baggageSearches: BaggageSearchModel[]): Promise<number> {
		const baggageSearchEntities = baggageSearches.map(baggageSearch =>
			this.toPersistence(baggageSearch),
		);

		const response = await this.baggageSearchRepository
			.getEntityManager()
			.upsertMany(BaggageSearchEntity, baggageSearchEntities);

		return response.length;
	}
}


import { BaggageSearchStatus } from '../baggage-search/domain/constants';
import { Entity, Enum, PrimaryKey, Property } from '@mikro-orm/postgresql';

@Entity({ tableName: 'baggages_search' })
export class BaggageSearchEntity {
	@PrimaryKey({ autoincrement: false })
	baggageSearchId: string;

	@Property()
	bagTag: string;

	@Property()
	flightId: string;

	@Enum({ items: () => BaggageSearchStatus })
	status: BaggageSearchStatus;
}

insert: baggageSearchEntities:

[

  {

    baggageSearchId: '575323f3-3b0e-4c04-8fef-eb49817f84a2',

    bagTag: '0075073986',

    flightId: 'b5c1d461-644a-41e5-adb5-1a0669ae36d1',

    status: 'STOPPED'

  },

  {

    baggageSearchId: '40ce1652-51b1-4950-8da5-eb742cc08de2',

    bagTag: '0075073999',

    flightId: 'b5c1d461-644a-41e5-adb5-1a0669ae36d1',

    status: 'STOPPED'

  }

]


should update:  
[

  {

    baggageSearchId: '575323f3-3b0e-4c04-8fef-eb49817f84a2',

    bagTag: '0075073986',

    flightId: 'b5c1d461-644a-41e5-adb5-1a0669ae36d1',

    status: 'SEARCH'

  },

  {

    baggageSearchId: '40ce1652-51b1-4950-8da5-eb742cc08de2',

    bagTag: '0075073999',

    flightId: 'b5c1d461-644a-41e5-adb5-1a0669ae36d1',

    status: 'SEARCH'

  }

]
```

### What driver are you using?

@mikro-orm/postgresql

### MikroORM version

6.2.0

### Node.js version

18

### Operating system

MacOs

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/mikro-orm/mikro-orm/blob/master/CONTRIBUTING.md).
- [X] Read the [docs](https://mikro-orm.io/docs).
- [X] Check that there isn't [already an issue](https://github.com/mikro-orm/mikro-orm/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/mikro-orm/mikro-orm/discussions) or join our [Discord](https://discord.gg/w8bjxFHS7X).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by B4nan** [comment]
> 6.2.0

Your first steps should be upgrading to latest version. 

If that does not help, provide a minimal but complete reproduction of what you are doing, from that snippet you have here it's not very clear what happens. 

Ideally base it on [this repository](https://github.com/mikro-orm/reproduction), don't use any webservers or additional abstractions, so its clear what happens. The reproduction needs to be executable and it needs to prepare some data if its supposed to be showcasing the update queries.

**Reply by B4nan** [reply]
Btw what might be happening is that if your target entity is already loaded in the context, you need to flush explicitly, such entity would be only modified in memory.

Try enabling debug mode to see what queries are executed.

**Reply by hugocruzlfc** [reply]
These are the consultations:

INFO [2024-08-09 16:01:50.567 +0000] (30752): [query] select "b0".* from "baggages_search" as "b0" where "b0"."bag_tag" = 'LA 045705' and "b0"."flight_id" = '41e6fa20-2a7d-4386-951f-f523af115e2a' limit 1 [took 13 ms, 1 result] 

INFO [2024-08-09 16:01:50.584 +0000] (30752): [query] select "b0".* from "baggages_search" as "b0" where "b0"."bag_tag" = 'LA 044503' and "b0"."flight_id" = '41e6fa20-2a7d-4386-951f-f523af115e2a' limit 1 [took 32 ms, 1 result] 

INFO [2024-08-09 16:01:50.585 +0000] (30752): [query] select "b0".* from "baggages_search" as "b0" where "b0"."bag_tag" = 'LA 045678' and "b0"."flight_id" = '41e6fa20-2a7d-4386-951f-f523af115e2a' limit 1 [took 33 ms, 1 result]

INFO [2024-08-09 16:01:50.590 +0000] (30752): [query] select "b0".* from "baggages_search" as "b0" where "b0"."bag_tag" = 'LA 045671' and "b0"."flight_id" = '41e6fa20-2a7d-4386-951f-f523af115e2a' limit 1 [took 38 ms, 1 result] 


and this is the function that calls my repository:


@Injectable()
export class UpsertBulkBaggageSearchUseCase implements IUseCase {
	private readonly logger = new Logger(UpsertBulkBaggageSearchUseCase.name);
	constructor(
		@InjectBaggageSearchRepository()
		readonly baggageRepository: IBaggagesSearchRepository,
	) {}

	async execute(
		flightId: VOUuid,
		bagTags: VOString[],
		status: VOBaggageSearchStatus,
	): Promise<number> {
		const baggageModelArrayPromise = bagTags.map(async bagTag => {
			const existingBaggageSearch =
				await this.baggageRepository.findByBagTagAndFlightId(bagTag, flightId);

			if (existingBaggageSearch) {
				existingBaggageSearch.updateStatus(status);
				return existingBaggageSearch;
			}
			return BaggageSearchModel.create(
				VOUuid.createRandom(),
				bagTag,
				flightId,
				status,
			);
		});

		const baggageModelArray = await Promise.all(baggageModelArrayPromise);

		console.log('use', baggageModelArray);

		const response = await this.baggageRepository.upsertMany(baggageModelArray);

		this.logger.log(
			`Upsert ${response} baggage search with flightId ${flightId.value} and status ${status.value}`,
		);

		return response;
	}
}

**Reply by hugocruzlfc** [reply]
But following your advice I did the following:

async upsertMany(baggageSearches: BaggageSearchModel[]): Promise<number> {
		const baggageSearchEntities = baggageSearches.map(baggageSearch =>
			this.toPersistence(baggageSearch),
		);

		this.baggageSearchRepository.getEntityManager().clear();

		const response = await this.baggageSearchRepository
			.getEntityManager()
			.upsertMany(BaggageSearchEntity, baggageSearchEntities);

		return response.length;
	}


and is now working correctly.

Thanks for everything man!!

**Reply by B4nan** [reply]
i guess i should change this, it already confused quite a lot of people that the updates are not happening for loaded entities

**Reply by B4nan** [reply]
maybe better to call `flush` after instead of `clear` before, you might be loosing some changes from memory
